api_fhir_r4/serializers/subscriptionSerializer.py

```diff
@@ -3,41 +3,65 @@
 from api_fhir_r4.exceptions import FHIRException
 from api_fhir_r4.models import Subscription
+from api_fhir_r4.permissions import FHIRApiInsureePermissions, FHIRApiInvoicePermissions
 from api_fhir_r4.serializers import BaseFHIRSerializer
 from api_fhir_r4.services import SubscriptionService
 from api_fhir_r4.utils import TimeUtils
 
 
 class SubscriptionSerializer(BaseFHIRSerializer):
     fhirConverter = SubscriptionConverter
+    _error_while_saving = 'Error while saving a subscription: %(msg)s'
+
+    resource_permissions = {
+        'patient': FHIRApiInsureePermissions.permissions_get,
+        'invoice': FHIRApiInvoicePermissions.permissions_get
+    }
 
     def to_internal_value(self, data):
         data['end'] = TimeUtils.str_iso_to_date(data['end'])
         return super().to_internal_value(data)
 
     def create(self, validated_data):
         user = self.context['request'].user
+        self.check_resource_rights(user, validated_data)
         service = SubscriptionService(user)
         copied_data = deepcopy(validated_data)
         del copied_data['_state'], copied_data['_original_state']
         result = service.create(copied_data)
-        if result.get('success', False):
-            return Subscription.objects.get(id=result['data']['id'])
-        else:
-            msg = result.get('message', 'Unknown')
-            raise FHIRException(f'Error while creating a subscription: {msg}')
+        return self.get_result_object(result)
 
     def update(self, instance, validated_data):
-        if str(instance.id) != validated_data['id']:
-            raise FHIRException(f'Error while updating a subscription: You cannot change the instance ID')
-
         user = self.context['request'].user
+        self.check_instance_id(instance, validated_data)
+        self.check_object_owner(user, instance)
+        self.check_resource_rights(user, validated_data)
         service = SubscriptionService(user)
         copied_data = {key: value for key, value in deepcopy(validated_data).items() if value is not None}
+        copied_data['id'] = instance.id
         del copied_data['_state'], copied_data['_original_state']
         result = service.update(copied_data)
+        return self.get_result_object(result)
+
+    def get_result_object(self, result):
         if result.get('success', False):
             return Subscription.objects.get(id=result['data']['id'])
         else:
-            msg = result.get('message', 'Unknown')
-            raise FHIRException(f'Error while updating a subscription: {msg}')
+            raise FHIRException(self._error_while_saving % {'msg': result.get('message', 'Unknown')})
+
+    def check_resource_rights(self, user, data):
+        resource_type = data.get('criteria', {}).get('resource_type', '').lower()
+        if not resource_type or resource_type not in self.resource_permissions:
+            raise FHIRException(self._error_while_saving % {'msg': f'Invalid resource_type ({resource_type})'})
```

Comment:
In serializer we should stick to Django Rest Framework exceptions, as this check is not actually part of FHIR logic. In this case I think  [ValidationError](https://www.django-rest-framework.org/api-guide/exceptions/#validationerror) will be optimal. 
Also you don't need `not resource_type` part, as default string will also fail in second condition. It might be good idea to separate non-existing criteria/resource_type from invalid one.

Reply:
I will remove the redundant ckeck and update the exception